### DISCUSSION TITLE
Question about Compound index of multiple fields

### DISCUSSION THREAD

**Discussion by xxaier** [qdrant/qdrant#2387] [Q&A]
Each of my queries needs to filter multiple fields. In this case, should I create an index for each field separately? 

Or, similar to mongodb, is it possible to create an index with multiple fields?

https://www.geeksforgeeks.org/mongodb-compound-indexes/

 (I don't see documentation describing how to create an index with multiple fields)

**Comment by timvisee** [comment]
> should I create an index for each field separately?

Yes.

Qdrant does not support compound indexes.

Maybe we should explicitly note this in the documentation to prevent confusion.

**Comment by eek** [comment]
Any plans for adding support for Compound indexes? @timvisee ?

**Reply by timvisee** [reply]
No short term plans.

Creating two separate payload indices has been good enough in the use cases we came across.

**Reply by eek** [reply]
> No short term plans.
> 
> Creating two separate payload indices has been good enough in the use cases we came across.

That's true, unless you're using Multitenancy, then, your main index is used for the tenant, and the filter for the actual field is non-indexable.

In our case, we have `library_id` for the tenant id, and then we need to filter sometimes by `media_id`.  If we didn't had tenants, media_id was sufficient, but because we have tenants, we're loosing our index to the tenant and `media_id` becomes unindexable without compound index.

It's funny that even Qdrant itself send us notification complaining we're lacking index for `media_id` when we actually have an index for `media_id`:

<img width="1600" height="883" alt="image" src="https://github.com/user-attachments/assets/472233ef-5351-44e7-9360-a3cec6313bec" />

**Reply by timvisee** [reply]
Qdrant supports multiple indices with `is_tenant: true`. Would that solve your situation?

> It's funny that even Qdrant itself send us notification complaining we're lacking index for media_id when we actually have an index for media_id

I'm a bit puzzled about the notification. Is it possible that the notification is old, from before the index was created?

**Reply by eek** [reply]
> I'm a bit puzzled about the notification. Is it possible that the notification is old, from before the index was created?

We've always had an index on `media_id` 🤔 We'll think of how we can optimize 🤔

**Reply by generall** [reply]
It says UUID index, but if your payload is not a valid UUID, it might skip index and cause this warning

**Reply by timvisee** [reply]
Here it says it did index points, so it should be valid UUIDs
<img width="205" height="164" alt="image" src="https://github.com/user-attachments/assets/973ae195-2cc3-4b72-ae3b-10536a7010c8" />

**Reply by generall** [reply]
query might be not a valid UUID

**Reply by legraphista** [reply]
> Qdrant supports multiple indices with `is_tenant: true`. Would that solve your situation?
> 

What happens to the HNSW index with multiple tenants?

If i understand correctly, one tenant id splits the HNSW into small graph segments, one segment per value.

With 2 tenant ids, does Qdrant create 2 sets of HNSW graph segments, one set per tenant, or does it split the graph into even smaller subgraphs, like a sparse Cartesian product between the 2 tenant ids?
